Calculate and test specificity for values and HTTP requests

mik047 opened the pull request:
**What**:  
Added `specificity` calculation logic for `Value` subclasses and HTTP requests. Refactored test utilities to use JUnit 5's `Stream<Arguments>` for better maintainability.

**Why**:  
To enhance comparison precision for paths, headers, query parameters, and bodies in HTTP requests, and to improve test code clarity and compliance.

**How**:  
- Introduced `specificity` methods to `Value` subclasses.  
- Implemented specificity score logic in `HttpRequest`.  
- Migrated parameterized tests to use `Stream<Arguments>`.  
- Added comprehensive unit tests for all changes.

**Checklist**:

- [x] Unit Tests
- [x] Build passing locally

mik047 committed:
Refactor parameterized test method to use Stream API

Replaced `List<Pair>` with `Stream<Arguments>` for test data in `urlPathToExpectedPathGenerality`, aligning with JUnit 5 practices. Simplified parameterized test method signature for better readability and maintainability.

mik047 committed:
Add specificity calculation for values and HTTP requests

Introduced `specificity` methods across `Value` subclasses to calculate specificity scores. Added corresponding logic to `HttpRequest` and comprehensive test cases to validate specificity calculations for paths, headers, query parameters, and body.

joelrosario commented on core/src/test/kotlin/io/specmatic/core/HttpRequestTest.kt:
```diff
@@ -412,8 +383,188 @@ internal class HttpRequestTest {
+            Arguments.of(null, 0),
+            Arguments.of("", 1),
+            Arguments.of("/", 1),
+            Arguments.of("/persons", 2),
```
This should be 1

joelrosario commented on core/src/main/kotlin/io/specmatic/core/value/JSONArrayValue.kt:
```diff
@@ -98,4 +98,8 @@ data class JSONArrayValue(override val list: List<Value>) : Value, ListValue, JS
     }
+
+    override fun specificity(): Int {
+        return list.sumOf { it.specificity() }
```
Return 1 if the list is empty

mik047 replied:
As discussed and concluded, we would treat empty list as `0`

joelrosario commented on core/src/main/kotlin/io/specmatic/core/value/JSONObjectValue.kt:
```diff
@@ -124,6 +124,10 @@ data class JSONObjectValue(val jsonObject: Map<String, Value> = emptyMap()) : Va
     }
+
+    override fun specificity(): Int {
+        return jsonObject.values.sumOf { it.specificity() }
```
return 1 if the object is empty

mik047 replied:
As discussed and concluded, we would treat empty map as `0`

joelrosario commented on core/src/main/kotlin/io/specmatic/core/value/Value.kt:
```diff
@@ -40,6 +40,10 @@ interface Value {
     }
+
+    fun specificity(): Int {
+        return 0
```
BinaryValue is not handle, and hence will return 0.

Perhaps, remove the defaulter and override everywhere else to return what's relevant. It's a small method and if we're creating a new Value later we need to make a conscious decision about the value to return.

mik047 replied:
made `specifity()` abstract for children to explicitly define relevant values.

Added values for BinaryValue and others where it was deriving default value.

mik047 committed:
Add `specificity` implementation to `Value` subclasses

Implemented the `specificity` function for various `Value` subclasses, such as `BinaryValue`, `NullValue`, and `NoBodyValue`. Updated tests in `HttpRequestTest` to include new cases for `specificity`. Prepared `XMLNode` for `specificity` with a placeholder implementation.

mik047 committed:
Add test cases for JSON object and array values in parameters

Added assertions for JSONObjectValue and JSONArrayValue to ensure proper handling of these cases in parameterized tests. This improves test coverage by verifying behavior for empty JSON structures.
